qa/smoke-test-http/src/test/java/org/elasticsearch/http/DanglingIndicesRestIT.java

```diff
@@ -0,0 +1,291 @@
+            String[] elements = nodeLine.split(" ");
+            if (elements[1].equals(nodeName)) {
+                return elements[0];
+            }
+        }
+
+        throw new AssertionError("Failed to map node name [" + nodeName + "] to node ID");
+    }
+
+    /**
+     * Helper that creates one or more indices, and importantly,
+     * checks that they are green before proceeding. This is important
+     * because the tests in this class stop and restart nodes, assuming
+     * that each index has a primary or replica shard on every node, and if
+     * a node is stopped prematurely, this assumption is broken.
+     *
+     * @return a mapping from each createad index name to its UUID
+     */
+    private Map<String, String> createIndices(String... indices) throws IOException {
+        assert indices.length > 0;
+        for (String index : indices) {
+            String indexSettings = "{"
+                + "  \"settings\": {"
+                + "    \"index\": {"
+                + "      \"number_of_shards\": 1,"
+                + "      \"number_of_replicas\": 2"
+                + "    }"
+                + "  }"
+                + "}";
+            Request request = new Request("PUT", "/" + index);
+            request.setJsonEntity(indexSettings);
+            final Response response = getRestClient().performRequest(request);
+            assertOK(response);
+        }
+        ensureGreen(indices);
+
+        final Response catResponse = getRestClient().performRequest(new Request("GET", "/_cat/indices?h=index,uuid"));
+        assertOK(catResponse);
+
+        final Map<String, String> createdIndexIDs = new HashMap<>();
+
+        final List<String> indicesAsList = Arrays.asList(indices);
+
+        for (String indexLine : Streams.readAllLines(catResponse.getEntity().getContent())) {
+            String[] elements = indexLine.split(" +");
+            if (indicesAsList.contains(elements[0])) {
+                createdIndexIDs.put(elements[0], elements[1]);
+            }
+        }
+
+        assertThat("Expected to find as many index UUIDs as created indices", createdIndexIDs.size(), equalTo(indices.length));
+
+        return createdIndexIDs;
+    }
+
+    private void deleteIndex(String indexName) throws IOException {
+        Response deleteResponse = getRestClient().performRequest(new Request("DELETE", "/" + indexName));
+        assertOK(deleteResponse);
+    }
+
+    private DanglingIndexDetails createDanglingIndices(String... indices) throws Exception {
+        ensureStableCluster(3);
+        final Map<String, String> indexToUUID = createIndices(indices);
+
+        final AtomicReference<String> stoppedNodeName = new AtomicReference<>();
+
+        // Restart node, deleting the index in its absence, so that there is a dangling index to recover
+        internalCluster().restartRandomDataNode(new InternalTestCluster.RestartCallback() {
```

Comment:
Are you sure that the dangling index metadata was written to disk by this point? I think it happens asynchronously now.

Reply:
I'm not sure - do you know how I would check? I had assumed that the tests would fail if the metadata wasn't written, because the create test explicitly checks what the dangling list API returns, and the import and delete tests both perform list calls before their respective actions.

Reply:
Yes I think these tests would fail if the metadata weren't written; I didn't see them fail but it might be pretty rare and I didn't try very hard to make it less rare, but nor could I see a reason why the metadata is definitely written by this point.

I don't think it's enough to list the dangling indices, or at least I don't see how this guarantees anything about the persistence of metadata of the non-dangling indices.

I was going to suggest adding a method to `PersistedState` but then I see that we already have one, see `GatewayMetaState#allPendingAsyncStatesWritten`. It should be enough to `assertBusy` that this returns true on all nodes.

Reply:
I see that you added an `assertBusy()` on `GatewayMetaState#allPendingAsyncStatesWritten` below. In fact I think we need to wait for that _before_ we restart the node, not after.